### PR TITLE
7330 Add functionality to check if a model is fine-tunable

### PR DESCRIPTION
**Pull request by joejuzl**
**Proposed changes**:
Adds a new method `rasa.model.can_fine_tune` which uses fingerprinting to determine if a model can be fine-tuned.

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/model.py [hunk 1]
```diff
@@ -80,6 +84,14 @@ class Section(NamedTuple):
 )
 SECTION_NLG = Section(name="NLG templates", relevant_keys=[FINGERPRINT_NLG_KEY])
 
+SECTION_FINE_TUNE = Section(
+    name="Fine-tune",
+    relevant_keys=[
+        FINGERPRINT_CONFIG_WITHOUT_EPOCHS_KEY,
+        FINGERPRINT_DOMAIN_WITHOUT_NLG_KEY,
```
Comment (joejuzl) [inline rasa/model.py:91]: Is this correct to ensure labels have not changed? Maybe there is a more looser definition e.g. `Domain.input_states`?

File: rasa/model.py [hunk 2]
```diff
@@ -468,6 +499,26 @@ def should_retrain(
         return fingerprint_comparison
 
 
+def can_fine_tune(last_fingerprint: Fingerprint, new_fingerprint: Fingerprint) -> bool:
+    """Check which components of a model should be retrained.
```
Comment (dakshvar22) [inline rasa/model.py:503]:
```suggestion
    """Check if components of a model can be finetuned with incremental training.
```

File: tests/core/test_model.py [hunk 3]
```diff
@@ -233,6 +266,44 @@ async def test_fingerprinting_changed_response_text(project: Text):
...
+async def test_fingerprinting_changing_config_epochs(project: Text):
+    importer = _project_files(project)
+
+    old_fingerprint = await model_fingerprint(importer)
+    config = await importer.get_config()
+
+    for key in ["pipeline", "policies"]:
+        for p in config[key]:
+            if "epochs" in p:
+                p["epochs"] += 10
```
Comment (dakshvar22) [inline tests/core/test_model.py:278]: Q: Should we create test config files and pass them to this test rather than explicitly increasing the epochs inside the test? Don't have a hard opinion but just wondering what would be the best way to test it.

File: rasa/model.py (hunk 1 above)
Comment (dakshvar22) [reply rasa/model.py:91]: I think it should be correct. I was first confused by the naming of `FINGERPRINT_DOMAIN_WITHOUT_NLG_KEY` as I thought this is the fingerprint of the domain without the `responses:` key but looks like they are also included inside `domain.fingerprint()`.
Maybe @wochinge has a better opinion on it.

**Review by dakshvar22 — CHANGES_REQUESTED** [review]
Looking good 👍 Please add types for pytests.

File: rasa/model.py (hunk 1 above)
Comment (wochinge) [reply rasa/model.py:91]: I'm positively surprised how useful the `Domain` is 😄 In my opinion it's fine!

File: tests/core/test_model.py [hunk 4]
```diff
@@ -178,6 +189,28 @@ def test_nlu_fingerprint_changed(fingerprint2, changed):
...
+        (_fingerprint(domain=["other"]), True),
+        (_fingerprint(rasa_version="100"), False),
+        (_fingerprint(config_core=["other"]), False),
+        (_fingerprint(config_nlu=["other"]), False),
+        (_fingerprint(nlu=["other"]), False),
+        (_fingerprint(nlg=["other"]), False),
+        (_fingerprint(stories=["other"]), False),
+    ],
+)
+def test_fine_tune_fingerprint_changed(fingerprint2, changed):
```
Comment (wochinge) [inline tests/core/test_model.py:206]: types please

File: tests/core/test_model.py [hunk 5]
```diff
@@ -418,3 +489,22 @@ async def get_domain() -> Domain:
...
 
     assert actual.is_empty()
+
+
+@pytest.mark.parametrize(
+    "min_compatible_version, old_model_version, can_tune",
+    [("2.1.0", "2.1.0", True), ("2.0.0", "2.1.0", True), ("2.1.0", "2.0.0", False),],
+)
+async def test_can_fine_tune_min_version(
+    project: Text, monkeypatch, old_model_version, min_compatible_version, can_tune
```
Comment (wochinge) [inline tests/core/test_model.py:499]: types please

File: tests/core/test_model.py [hunk 6]
```diff
@@ -233,6 +266,44 @@ async def test_fingerprinting_changed_response_text(project: Text):
...
+
+    old_fingerprint = await model_fingerprint(importer)
+    config = await importer.get_config()
+
+    for key in ["pipeline", "policies"]:
+        for p in config[key]:
+            if "epochs" in p:
+                p["epochs"] += 10
+
+    importer.get_config = asyncio.coroutine(lambda: config)
```
Comment (wochinge) [inline tests/core/test_model.py:280]: `asyncio.coroutine` is deprecated 😬

**Review by wochinge — COMMENTED** [review]
production code looks great! Need to have another look at the tests later 👍

File: tests/core/test_model.py (hunk 3 above)
Comment (joejuzl) [reply tests/core/test_model.py:278]: Good question. The problem I see with that is someone could change the number of epochs in one of the files, which would not make this test fail, but mean it wasn't actually testing anything.

Comment (dakshvar22) [reply tests/core/test_model.py:280]: Hmm, that makes sense. I think the place where having it in a file would be useful is when you are writing a test to check if the right number of epochs are passed to the ML components during loading. I guess that would be part of your other PR or maybe my PR.

Comment (wochinge) [reply tests/core/test_model.py:280]: I think we could simplify the test by doing something like:

1. Define config as `Dict`
2. Write to a file
3. `TrainingDataImporter.load_from_config`
4. Test fingerprint

**Review by wochinge — APPROVED** [review]
Nice 👍

**Comment by wochinge** [comment]
@dakshvar22 is the `changes request` still valid?

**Review by dakshvar22 — APPROVED** [review]
LGTM 👍
